hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -942,17 +944,159 @@ private boolean isKeyEmpty(OmKeyInfo keyInfo) {
+    }
+
+    validateListKeyArgs(volumeName, bucketName);
+    List<OmKeyInfo> keyList = new ArrayList<>();
+    if (maxKeys <= 0) {
+      return keyList;
+    }
+
+    if (OzoneManagerRatisUtils.isOmLayoutVersionV1()) {
+      return listKeysV1(volumeName, bucketName, startKey, keyPrefix, maxKeys);
+    }
+
+    keyList = metadataManager.listKeys(volumeName, bucketName, startKey,
+            keyPrefix, maxKeys);
+
+    return keyList;
+  }
+
+  private List<OmKeyInfo> listKeysV1(String volumeName, String bucketName,
+      String startKey, String keyPrefix, int maxKeys) throws IOException {
+    List<OmKeyInfo> keyList = new ArrayList<>();
+    if (StringUtils.isBlank(startKey)) {
+      return listKeysRecursivelyFromKeyPrefixV1(volumeName, bucketName,
+              keyPrefix, maxKeys, keyList);
+    } else {
+      return listKeysRecursivelyFromStartKeyV1(volumeName, bucketName, startKey,
+              keyPrefix, maxKeys, keyList);
+    }
+  }
+
+  private List<OmKeyInfo> listKeysRecursivelyFromKeyPrefixV1(String volumeName,
+      String bucketName, String keyPrefix, int maxKeys,
+      List<OmKeyInfo> keyList) throws IOException {
+
+    String startKey = "";
+    int countEntries = keyList.size();
+    List<OzoneFileStatus> keys = listOzoneFileStatusesV1(false,
+            startKey, countEntries, maxKeys, null, volumeName,
+            bucketName, keyPrefix, true, true);
+
+    for (OzoneFileStatus status : keys) {
+      OmKeyInfo omKeyInfo = status.getKeyInfo();
+      keyList.add(omKeyInfo);
+      if (status.isDirectory()) {
+        // add trailing slash to represent directory
+        String dirName = omKeyInfo.getKeyName();
+        if (!dirName.endsWith("/")) {
+          omKeyInfo.setKeyName(dirName + OM_KEY_PREFIX);
+        }
+        if (!StringUtils.equals(dirName, keyPrefix)) {
+          listKeysRecursivelyFromKeyPrefixV1(volumeName, bucketName, dirName,
+                  maxKeys, keyList);
+        }
+      }
+    }
+    return keyList;
+  }
+
+  private List<OmKeyInfo> listKeysRecursivelyFromStartKeyV1(String volumeName,
+      String bucketName, String startKey, String keyPrefix,
+      int maxKeys, List<OmKeyInfo> keyList) throws IOException {
+
+    if (startKey.startsWith(keyPrefix)) {
+      // Back and forth searching method. Here, it needs to move the startKey
+      // & keyprefix pointers one step backward, then list all the keys
+      // forward to these to pointers.
+
+      // For example, startKey="dir1/dir2/dir3/file3" and keyPrefix="dir1/".
+      // Iterate then move startkey and parentKeyPrefixPath positions till the
+      // given keyPrefix and lists all its sup paths...
+
+      // Iteration-1:
+      // startKeyPos="dir1/dir2/dir3/file3", parentKeyPrefix="dir1/dir2/dir3"
+      // Iteration-2:
+      // startKeyPos="dir1/dir2/dir3", parentKeyPrefix="dir1/dir2"
+      // Iteration-3:
+      // startKeyPos="dir1/dir2", parentKeyPrefix="dir1"
```

Comment:
Not understand why we divided the startKey to search. In addition, why do we have to check if startKey is started with keyPrefix?
Can we reuse listOzoneFileStatusesV1 method like listKeysRecursivelyFromKeyPrefixV1 does? One difference is that now startKey is not empty.

Reply:
Thanks @linyiqun for the example case. Please see my below reply.

> In Iteration-2, will it search for the dir1/dir2/dir3 again? I think it will be skipped as dir1/dir2/dir3 is the new startKey will be skipped.
Yes, it will skip the startKey from the resultList when startKey is not empty.

Secondly, the `listKeysV1()` algorithm is written like a template fashion with the following steps irrespective of startKey empty or non-empty:
step-1) It will fetch all the files under the given path.
step-2) It will fetch all the directories under the given path.

If the startKey is a directory, then the assumption is that all the files under this key was already visited in previous batch. So, in your example, you have passed `startKeyPos="dir1/dir2/dir3"`. Here it will check any nextString directory exists after the startKeyPos in DirTable. In your example case, we don't have any , so will return empty dir list. Assume, there was "dir1/dir2/dir4" exists, then it would have added to the result list.
Now, it won't go to FileTable and list all keys after `startKeyPos="dir1/dir2/dir3"`, instead it will move startKeyPos pointer to previous dir "dir1/dir2" and do a check ` !StringUtils.equals(parentStartKeyPath, keyPrefix)`. This return false.

So the resultList in this case is, `empty result list []`

Hope this explanation is clear to you.